src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,548 @@
+            max[random_dimension] = temp
+            min[random_dimension] = p
+            right = self.build_tree(min, max, current_depth + 1)
+            return HalfSpaceTreeNode(left=left, right=right, split_attribute=random_dimension,
+                                     split_value=p, depth=current_depth)
+
+    def update_mass(self, X, is_reference_window):
+        """
+        Records mass data in the tree.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+        is_reference_window: Boolean
+            True if the classifier is in the first reference window, otherwise false
+        Returns
+        -------
+
+        """
+        return self._update_mass(X, self.root, is_reference_window)
+
+    def _update_mass(self, X, node, is_reference_window):
+        """
+        Private function where actual mass data is recorded.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+        node: HalfSpaceTreeNode
+
+        is_reference_window: Boolean
+            True if the classifier is in the first reference window, otherwise false
+
+        """
+        if node is None:
+            return
+
+        if is_reference_window:
+            node.r = node.r + 1
+        else:
+            node.l = node.l + 1
+
+        if node.depth < self.max_depth:
+            if X[node.split_attribute] > node.split_value:
+                self._update_mass(X, node.right, is_reference_window)
+            else:
+                self._update_mass(X, node.left, is_reference_window)
+        else:
+            return
+
+    def update_model(self, node):
+        """
+        Updates model to adapt to new data.
+
+        Parameters
+        ----------
+        node: HalfSpaceTreeNode
+
+        """
+        if node is not None:
+            if node.r is not 0 and node.l is not 0:
+                node.r = node.l
+
+            if node.l is not 0:
+                node.l = 0
+
+            self.update_model(node.left)
+            self.update_model(node.right)
+        return
+
+
+class HalfSpaceTreeNode:
+
+    def __init__(self, left=None, right=None, split_value=0.0, split_attribute=-1,
+                 depth=0, internal_node=True):
+        """
+        Parameters
+        ----------
+        left: HalfSpaceTreeNode, optional (default=None)
+            The left node pointed to this node.
+
+        right: HalfSpaceTreeNode, optional (default=None)
+            The left node pointed to this node.
+
+        split_value: double, optional (default=0.0)
+            Mid point of the selected random dimension.
+            'p' in the original paper.
+
+        split_attribute: int, optional (default=-1)
+            Randomly selected attribute for the node.
+            'q' in the original paper.
+
+        depth: int, optional (default=0)
```

Comment:
According to the documentation in MOA, the default value should be `1`:
> The depth of this node in the HSTree. The root node is at depth 1.

Reply:
I don't see anything in the paper justifying the default value to 1. Here depth, k in the paper, stands for the depth of the node in the tree. The root of a tree has depth one but, maximum anomaly score of a node is `max_score = self.window_size * pow(2.0, self.depth)`. And then this value is subtracted by one. So, given a tree with only one node in it, anomaly score for the tree with depth 1 and depth 0 will be 0.5 and 0 respectively. From my point of view, since every data is in the same place, there should be no anomaly in this case and that is why I put 0 as the default value.

Reply:
I see, thanks for the clarification. This is indeed strange, given that the root node of a tree should have a depth of `0`. As you mention, if there is only one node we cannot define an anomaly value thus `score=0`.

However, this change has a direct impact on the performance in the method. It is important to keep it in mind when validating the results from your testing. This could be the root cause on the performance difference you reported.

Reply:
Actually, this was the root of the performance difference that I have reported, now, depending on the parameters, auc score is around 0.998.